pkg/assembler/backends/testing/isOccurrence.go

```diff
@@ -17,209 +17,343 @@ package testing
+			for _, version := range pkgVersion.versions {
+				if noMatchInput(packageArg.Version, version.version) {
+					continue
 				}
-			} else if val, ok := occurrence.Subject.(model.Source); ok {
-				if reflect.DeepEqual(val, *selectedSource) {
-					return occurrence, nil
+				if noMatchInput(packageArg.Subpath, version.subpath) {
+					continue
 				}
+				if !reflect.DeepEqual(version.qualifiers, getQualifiersFromInput(packageArg.Qualifiers)) {
+					continue
+				}
+				if found {
+					return nil, gqlerror.Errorf("More than one package matches input")
+				}
+				packageID = version.id
+				found = true
+			}
+			if !found {
+				return nil, gqlerror.Errorf("No package matches input")
 			}
 		}
 	}
 
-	newIsOccurrence := &model.IsOccurrence{
-		Justification: justification,
-		Artifact:      artifact,
-		Origin:        origin,
-		Collector:     collector,
+	sourceID := maxUint32
+	if subject.Source != nil {
+		source := subject.Source
+		srcNamespace, srcHasNamespace := c.sources[source.Type]
+		if !srcHasNamespace {
+			return nil, gqlerror.Errorf("IngestOccurrence :: Source type \"%s\" not found", source.Type)
+		}
+		srcName, srcHasName := srcNamespace.namespaces[source.Namespace]
+		if !srcHasName {
+			return nil, gqlerror.Errorf("IngestOccurrence :: Source namespace \"%s\" not found", source.Namespace)
+		}
+		found := false
+		for _, src := range srcName.names {
+			if src.name != source.Name {
+				continue
+			}
+			if noMatchInput(source.Tag, src.tag) {
+				continue
+			}
+			if noMatchInput(source.Commit, src.commit) {
+				continue
+			}
+			if found {
+				return nil, gqlerror.Errorf("IngestOccurrence :: More than one source matches input")
+			}
+			sourceID = src.id
+			found = true
+		}
+		if !found {
+			return nil, gqlerror.Errorf("IngestOccurrence :: No source matches input")
+		}
+	}
+
+	// could search backedges for pkg/src or artifiact, just do artifact
+	for _, id := range a.occurrences {
+		o, _ := c.occurrenceByID(id)
+		if o.pkg == packageID &&
+			o.source == sourceID &&
+			o.artifact == a.id &&
+			o.justification == occurrence.Justification &&
+			o.origin == occurrence.Origin &&
+			o.collector == occurrence.Collector {
+			return c.convOccurrence(o), nil
+		}
 	}
-	if selectedPackage != nil {
-		newIsOccurrence.Subject = selectedPackage
+	o := &isOccurrenceStruct{
+		id:            c.getNextID(),
+		pkg:           packageID,
+		source:        sourceID,
+		artifact:      a.id,
+		justification: occurrence.Justification,
+		origin:        occurrence.Origin,
+		collector:     occurrence.Collector,
+	}
+	c.index[o.id] = o
+	a.occurrences = append(a.occurrences, o.id)
+	if packageID != maxUint32 {
+		pv, pn, _ := c.pkgByID(packageID)
+		if pv != nil {
+			pv.occurrences = append(pv.occurrences, o.id)
+		} else {
+			pn.occurrences = append(pn.occurrences, o.id)
+		}
 	} else {
-		newIsOccurrence.Subject = selectedSource
+		s, _ := c.sourceByID(sourceID)
+		s.occurrences = append(s.occurrences, o.id)
```

Comment:
helper function like the others

Reply:
The getters seem unnecessary, as the struct members are being freely accessed in general. In fact all the struct members should be capital. If the members were truly private, the getter should just be the capital of the internal member (https://go.dev/doc/effective_go#Getters). The setter helpers should probably be renamed AddFoo.

I'll update to match for constancy here, but we should revisit these.

Reply:
done